File: packages/agent/src/scanview_agent/key_images.py
from __future__ import annotations

import hashlib
import json
import math
import re
import struct
import zipfile
import zlib
from datetime import datetime
from pathlib import Path
from typing import Any, BinaryIO

from .measurements import validate_measurement_packet


OPAQUE_ID = re.compile(r"^[0-9a-f]{16}$")
SHA256 = re.compile(r"^[0-9a-f]{64}$")
ORIENTATION = re.compile(r"^[RLAPHF]{1,3}$")
EXPECTED_ARCHIVE_FILES = {"key-image.json", "key-image.png", "measurements.json"}
MAX_ARCHIVE_MEMBER_BYTES = 64 * 1024 * 1024


def _has_only_keys(value: dict[str, Any], allowed: set[str]) -> bool:
    return not (set(value) - allowed)


def _valid_opaque_id(value: Any, kind: str) -> bool:
    return isinstance(value, str) and bool(
        OPAQUE_ID.fullmatch(value)
        or re.fullmatch(rf"{re.escape(kind)}_[0-9a-f]{{20}}", value)
    )


def _finite_number(value: Any) -> bool:
    return (
        isinstance(value, (int, float))
        and not isinstance(value, bool)
        and math.isfinite(value)
    )


def _positive_integer(value: Any) -> bool:
    return isinstance(value, int) and not isinstance(value, bool) and value > 0


def _valid_datetime(value: Any) -> bool:
    if not isinstance(value, str) or "T" not in value:
        return False
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return False
    return parsed.tzinfo is not None


def validate_key_image_packet(packet: Any) -> list[str]:
    errors: list[str] = []
    if not isinstance(packet, dict):
        return ["key-image packet must be a JSON object"]
    if not _has_only_keys(
        packet,
        {
            "schema_version",
            "created_at",
            "review_status",
            "artifact_type",
            "source",
            "display",
            "image",
            "measurement_evidence",
            "implementation",
            "limitations",
        },
    ):
        errors.append("key-image packet contains unsupported fields")
    schema_version = packet.get("schema_version")
    if schema_version not in {"1.0.0", "2.0.0"}:
        errors.append("schema_version must be 1.0.0 or 2.0.0")
    if not _valid_datetime(packet.get("created_at")):
        errors.append("created_at must be an ISO 8601 date-time")
    if packet.get("review_status") != "unreviewed":
        errors.append("review_status must be unreviewed")
    if packet.get("artifact_type") != "derived_display_key_image":
        errors.append("artifact_type must be derived_display_key_image")

    source = packet.get("source")
    if not isinstance(source, dict):
        errors.append("source must be an object")
    else:
        if not _has_only_keys(
            source,
            {
                "study_id",
                "series_id",
                "instance_id",
                "patient_context_id",
                "frame_of_reference_id",
                "modality",
                "acquisition_date",
                "series_description",
                "instance_number",
            },
        ):
            errors.append("source contains unsupported fields")
        study = source.get("study_id")
        patient_context = source.get("patient_context_id")
        if schema_version == "2.0.0":
            if not _valid_opaque_id(study, "study"):
                errors.append("source.study_id must be a supported opaque ID")
            if not _valid_opaque_id(patient_context, "patient"):
                errors.append("source.patient_context_id must be a supported opaque ID")
        elif study is not None or patient_context is not None:
            errors.append("key-image v1 source cannot contain v2 context fields")
        if not _valid_opaque_id(source.get("series_id"), "series"):
            errors.append("source.series_id must be a supported opaque ID")
        if not _valid_opaque_id(source.get("instance_id"), "instance"):
            errors.append("source.instance_id must be a supported opaque ID")
        frame = source.get("frame_of_reference_id")
        if frame is not None and not _valid_opaque_id(frame, "frame"):
            errors.append("source.frame_of_reference_id must be a supported opaque ID")
        if not isinstance(source.get("modality"), str) or not source["modality"]:
            errors.append("source.modality must be a non-empty string")
        acquisition_date = source.get("acquisition_date")
        if acquisition_date is not None and (
            not isinstance(acquisition_date, str)
            or not re.fullmatch(r"[0-9]{8}", acquisition_date)
        ):
            errors.append("source.acquisition_date must be an eight-digit DICOM date")
        if (
            not isinstance(source.get("series_description"), str)
            or not source["series_description"]
        ):
            errors.append("source.series_description must be a non-empty string")
        if not isinstance(source.get("instance_number"), int) or isinstance(
            source.get("instance_number"), bool
        ):
            errors.append("source.instance_number must be an integer")

    display = packet.get("display")
    if not isinstance(display, dict):
        errors.append("display must be an object")
    else:
        if not _has_only_keys(
            display,
            {
                "viewport_role",
                "stack_position",
                "stack_count",
                "source_kind",
                "viewport_width_px",
                "viewport_height_px",
                "patient_orientation",
                "presentation",
            },
        ):
            errors.append("display contains unsupported fields")
        if display.get("viewport_role") not in {"baseline", "followup"}:
            errors.append("display.viewport_role is invalid")
        if not _positive_integer(display.get("stack_position")):
            errors.append("display.stack_position must be a positive integer")
        if not _positive_integer(display.get("stack_count")):
            errors.append("display.stack_count must be a positive integer")
        if (
            _positive_integer(display.get("stack_position"))
            and _positive_integer(display.get("stack_count"))
            and display["stack_position"] > display["stack_count"]
        ):
            errors.append("display.stack_position exceeds stack_count")
        if display.get("source_kind") not in {"browser-folder", "loopback-service"}:
            errors.append("display.source_kind is invalid")
        for key in ("viewport_width_px", "viewport_height_px"):
            if not _positive_integer(display.get(key)):
                errors.append(f"display.{key} must be a positive integer")
        orientation = display.get("patient_orientation")
        if orientation is not None:
            if not isinstance(orientation, dict) or not _has_only_keys(
                orientation, {"left", "right", "top", "bottom"}
            ):
                errors.append("display.patient_orientation is invalid")
            elif not all(
                isinstance(orientation.get(key), str)
                and bool(ORIENTATION.fullmatch(orientation[key]))
                for key in ("left", "right", "top", "bottom")
            ):
                errors.append("display.patient_orientation labels are invalid")
        presentation = display.get("presentation")
        if not isinstance(presentation, dict) or not _has_only_keys(
            presentation, {"voi_range", "invert", "zoom", "pan"}
        ):
            errors.append("display.presentation is invalid")
        else:
            voi_range = presentation.get("voi_range")
            if voi_range is not None:
                if (
                    not isinstance(voi_range, dict)
                    or not _has_only_keys(voi_range, {"lower", "upper"})
                    or not _finite_number(voi_range.get("lower"))
                    or not _finite_number(voi_range.get("upper"))
                    or voi_range["upper"] <= voi_range["lower"]
                ):
                    errors.append("display.presentation.voi_range is invalid")
            invert = presentation.get("invert")
            if invert is not None and not isinstance(invert, bool):
                errors.append("display.presentation.invert must be boolean")
            zoom = presentation.get("zoom")
            if zoom is not None and (not _finite_number(zoom) or zoom <= 0):
                errors.append("display.presentation.zoom must be positive and finite")
            pan = presentation.get("pan")
            if pan is not None and (
                not isinstance(pan, list)
                or len(pan) != 2
                or not all(_finite_number(value) for value in pan)
            ):
                errors.append("display.presentation.pan must contain two finite numbers")

    image = packet.get("image")
    if not isinstance(image, dict) or not _has_only_keys(
        image, {"filename", "mime_type", "width_px", "height_px", "sha256"}
    ):
        errors.append("image must be a supported object")
    else:
        if image.get("filename") != "key-image.png":
            errors.append("image.filename must be key-image.png")
        if image.get("mime_type") != "image/png":
            errors.append("image.mime_type must be image/png")
        for key in ("width_px", "height_px"):
            if not _positive_integer(image.get(key)):
                errors.append(f"image.{key} must be a positive integer")
        if not isinstance(image.get("sha256"), str) or not SHA256.fullmatch(image["sha256"]):
            errors.append("image.sha256 must be a lowercase SHA-256 digest")

    measurements = packet.get("measurement_evidence")
    if not isinstance(measurements, dict) or not _has_only_keys(
        measurements,
        {"filename", "schema_version", "measurement_count", "tracking_ids", "sha256"},
    ):
        errors.append("measurement_evidence must be a supported object")
    else:
        if measurements.get("filename") != "measurements.json":
            errors.append("measurement_evidence.filename must be measurements.json")
        if measurements.get("schema_version") != "3.0.0":
            errors.append("measurement_evidence.schema_version must be 3.0.0")
        count = measurements.get("measurement_count")
        if not isinstance(count, int) or isinstance(count, bool) or count < 0:
            errors.append("measurement_evidence.measurement_count must be non-negative")
        tracking_ids = measurements.get("tracking_ids")
        if not isinstance(tracking_ids, list) or not all(
            isinstance(value, str) and value for value in tracking_ids
        ):
            errors.append("measurement_evidence.tracking_ids must be a string array")
        elif len(tracking_ids) != len(set(tracking_ids)):
            errors.append("measurement_evidence.tracking_ids must be unique")
        elif isinstance(count, int) and count != len(tracking_ids):
            errors.append("measurement_evidence count disagrees with tracking_ids")
        if not isinstance(measurements.get("sha256"), str) or not SHA256.fullmatch(
            measurements["sha256"]
        ):
            errors.append("measurement_evidence.sha256 must be a lowercase SHA-256 digest")

    implementation = packet.get("implementation")
    if not isinstance(implementation, dict) or not _has_only_keys(
        implementation, {"name", "version", "renderer"}
    ):
        errors.append("implementation must be a supported object")
    else:
        expected_exporter_version = "0.2.0" if schema_version == "2.0.0" else "0.1.0"
        if (
            implementation.get("name") != "ScanView key-image exporter"
            or implementation.get("version") != expected_exporter_version
            or implementation.get("renderer") != "Cornerstone3D 5.8.2"
        ):
            errors.append("implementation is unsupported")

    limitations = packet.get("limitations")
    if not isinstance(limitations, list) or not limitations or not all(
        isinstance(value, str) and value for value in limitations
    ):
        errors.append("limitations must be a non-empty string array")
    return errors


def _png_dimensions(data: bytes) -> tuple[int, int] | None:
    if len(data) < 45 or data[:8] != b"\x89PNG\r\n\x1a\n":
        return None
    offset = 8
    dimensions: tuple[int, int] | None = None
    saw_idat = False
    saw_iend = False
    while offset + 12 <= len(data):
        length = struct.unpack(">I", data[offset : offset + 4])[0]
        chunk_end = offset + 12 + length
        if chunk_end > len(data):
            return None
        chunk_type = data[offset + 4 : offset + 8]
        chunk_data = data[offset + 8 : offset + 8 + length]
        expected_crc = struct.unpack(">I", data[offset + 8 + length : chunk_end])[0]
        if zlib.crc32(chunk_type + chunk_data) & 0xFFFFFFFF != expected_crc:
            return None
        if offset == 8:
            if chunk_type != b"IHDR" or length != 13:
                return None
            width, height = struct.unpack(">II", chunk_data[:8])
            if width <= 0 or height <= 0:
                return None
            dimensions = (width, height)
        elif chunk_type == b"IDAT":
            saw_idat = True
        elif chunk_type == b"IEND":
            if length != 0 or chunk_end != len(data):
                return None
            saw_iend = True
            break
        offset = chunk_end
    return dimensions if dimensions and saw_idat and saw_iend else None


def key_image_archive_summary(path: Path | BinaryIO) -> dict[str, Any]:
    errors: list[str] = []
    packet: Any = None
    measurements: Any = None
    png_bytes = b""
    measurement_bytes = b""
    try:
        with zipfile.ZipFile(path) as archive:
            infos = archive.infolist()
            names = {info.filename for info in infos}
            if names != EXPECTED_ARCHIVE_FILES or len(infos) != len(EXPECTED_ARCHIVE_FILES):
                errors.append("archive must contain exactly the three supported files")
            if any(info.flag_bits & 0x1 for info in infos):
                errors.append("encrypted archive members are unsupported")
            if any(info.file_size > MAX_ARCHIVE_MEMBER_BYTES for info in infos):
                errors.append("archive member exceeds the local safety limit")
            if not errors:
                packet_bytes = archive.read("key-image.json")
                png_bytes = archive.read("key-image.png")
                measurement_bytes = archive.read("measurements.json")
                packet = json.loads(packet_bytes)
                measurements = json.loads(measurement_bytes)
    except (OSError, zipfile.BadZipFile, KeyError, UnicodeDecodeError, json.JSONDecodeError) as error:
        errors.append(f"archive could not be read: {type(error).__name__}")

    packet_errors = validate_key_image_packet(packet) if packet is not None else []
    errors.extend(packet_errors)
    measurement_errors = (
        validate_measurement_packet(measurements) if measurements is not None else []
    )
    errors.extend(f"measurements.json: {error}" for error in measurement_errors)

    image_integrity = False
    measurement_integrity = False
    if isinstance(packet, dict):
        image = packet.get("image")
        if isinstance(image, dict) and png_bytes:
            dimensions = _png_dimensions(png_bytes)
            image_integrity = bool(
                dimensions
                and hashlib.sha256(png_bytes).hexdigest() == image.get("sha256")
                and dimensions == (image.get("width_px"), image.get("height_px"))
            )
            if not image_integrity:
                errors.append("key-image.png digest or dimensions disagree with key-image.json")
        evidence = packet.get("measurement_evidence")
        if isinstance(evidence, dict) and measurement_bytes:
            measurement_integrity = (
                hashlib.sha256(measurement_bytes).hexdigest() == evidence.get("sha256")
            )
            if not measurement_integrity:
                errors.append("measurements.json digest disagrees with key-image.json")

    if isinstance(packet, dict) and isinstance(measurements, dict):
        evidence = packet.get("measurement_evidence")
        records = measurements.get("measurements")
        if isinstance(evidence, dict) and isinstance(records, list):
            tracking_ids = [
                record.get("tracking_id") for record in records if isinstance(record, dict)
            ]
            if evidence.get("measurement_count") != len(records):
                errors.append("measurement count disagrees with measurements.json")
            if evidence.get("tracking_ids") != tracking_ids:
                errors.append("tracking IDs disagree with measurements.json")
            source = packet.get("source")
            if isinstance(source, dict) and any(
                not isinstance(record, dict)
                or not isinstance(record.get("source"), dict)
                or record["source"].get("series_id") != source.get("series_id")
                or record["source"].get("instance_id") != source.get("instance_id")
                for record in records
            ):
                errors.append("measurement source disagrees with the key-image source instance")

    measurement_count = 0
    if isinstance(packet, dict) and isinstance(packet.get("measurement_evidence"), dict):
        value = packet["measurement_evidence"].get("measurement_count")
        if isinstance(value, int) and not isinstance(value, bool):
            measurement_count = value
    return {
        "valid": not errors,
        "schema_version": packet.get("schema_version") if isinstance(packet, dict) else None,
        "review_status": packet.get("review_status") if isinstance(packet, dict) else None,
        "artifact_type": packet.get("artifact_type") if isinstance(packet, dict) else None,
        "measurement_count": measurement_count,
        "image_integrity": image_integrity,
        "measurement_integrity": measurement_integrity,
        "errors": errors,
    }

File: apps/viewer/src/keyImages.ts
import { strToU8, zipSync } from 'fflate';
import type { PatientOrientationLabels } from './dicom';
import type { MeasurementEvidencePacket } from './measurements';

export type KeyImagePresentation = {
  voi_range?: { lower: number; upper: number };
  invert?: boolean;
  zoom?: number;
  pan?: [number, number];
};

export type KeyImageEvidencePacket = {
  schema_version: '2.0.0';
  created_at: string;
  review_status: 'unreviewed';
  artifact_type: 'derived_display_key_image';
  source: {
    study_id: string;
    series_id: string;
    instance_id: string;
    patient_context_id: string;
    frame_of_reference_id?: string;
    modality: string;
    acquisition_date?: string;
    series_description: string;
    instance_number: number;
  };
  display: {
    viewport_role: 'baseline' | 'followup';
    stack_position: number;
    stack_count: number;
    source_kind: 'browser-folder' | 'loopback-service';
    viewport_width_px: number;
    viewport_height_px: number;
    patient_orientation?: PatientOrientationLabels;
    presentation: KeyImagePresentation;
  };
  image: {
    filename: 'key-image.png';
    mime_type: 'image/png';
    width_px: number;
    height_px: number;
    sha256: string;
  };
  measurement_evidence: {
    filename: 'measurements.json';
    schema_version: '3.0.0';
    measurement_count: number;
    tracking_ids: string[];
    sha256: string;
  };
  implementation: {
    name: 'ScanView key-image exporter';
    version: '0.2.0';
    renderer: 'Cornerstone3D 5.8.2';
  };
  limitations: string[];
};

export type ConsultationSelectionSlot = 'view_a' | 'view_b';

export const CONSULTATION_KEY_IMAGE_IMPLEMENTATION = {
  name: 'ScanView consultation key-image normalizer',
  version: '0.1.0',
  renderer: 'Cornerstone3D 5.8.2',
  source_key_image_schema: '2.0.0',
} as const;

export const CONSULTATION_KEY_IMAGE_LIMITATIONS = [
  'This PNG is an unreviewed derived display capture for source-image discussion only; original DICOM remains authoritative.',
  'The packet-local view slot does not establish chronology, lesion matching, diagnosis, or treatment response.',
  'Rendered pixels and burned-in text may identify the patient; this evidence is sensitive and not deidentified.',
] as const;

export type ConsultationKeyImageEvidencePacket = {
  schema_version: '1.0.0';
  created_at: string;
  review_status: 'unreviewed';
  artifact_type: 'derived_display_consultation_key_image';
  source: KeyImageEvidencePacket['source'];
  display: Omit<KeyImageEvidencePacket['display'], 'viewport_role'> & {
    selection_slot: ConsultationSelectionSlot;
  };
  image: KeyImageEvidencePacket['image'];
  measurement_evidence: KeyImageEvidencePacket['measurement_evidence'];
  implementation: typeof CONSULTATION_KEY_IMAGE_IMPLEMENTATION;
  limitations: [...typeof CONSULTATION_KEY_IMAGE_LIMITATIONS];
};

type BuildPacketInput = {
  createdAt: string;
  source: KeyImageEvidencePacket['source'];
  display: KeyImageEvidencePacket['display'];
  imageWidth: number;
  imageHeight: number;
  pngBytes: Uint8Array;
  measurementPacket: MeasurementEvidencePacket;
  measurementBytes: Uint8Array;
};

type BuildConsultationPacketInput = Omit<BuildPacketInput, 'display'> & {
  display: ConsultationKeyImageEvidencePacket['display'];
};

export type KeyImageArchiveInput = {
  viewportCanvas: HTMLCanvasElement;
  annotationSvg?: SVGSVGElement;
  orientationLabels?: PatientOrientationLabels;
  viewportRole: 'baseline' | 'followup';
  source: KeyImageEvidencePacket['source'];
  display: Omit<
    KeyImageEvidencePacket['display'],
    'viewport_role' | 'viewport_width_px' | 'viewport_height_px' | 'patient_orientation'
  >;
  measurementPacket: MeasurementEvidencePacket;
  createdAt?: string;
};

export type KeyImageArchive = {
  filename: string;
  packet: KeyImageEvidencePacket;
  bytes: Uint8Array;
};

export type ConsultationKeyImageArchiveInput = Omit<
  KeyImageArchiveInput,
  'viewportRole'
> & {
  selectionSlot: ConsultationSelectionSlot;
};

export type ConsultationKeyImageArchive = {
  filename: string;
  packet: ConsultationKeyImageEvidencePacket;
  bytes: Uint8Array;
};

const sha256Hex = async (bytes: Uint8Array): Promise<string> => {
  const digest = await crypto.subtle.digest('SHA-256', bytes.slice().buffer);
  return Array.from(new Uint8Array(digest))
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('');
};

const canvasToPngBytes = (canvas: HTMLCanvasElement): Promise<Uint8Array> =>
  new Promise((resolve, reject) => {
    canvas.toBlob((blob) => {
      if (!blob) {
        reject(new Error('The browser could not encode the key image.'));
        return;
      }
      void blob.arrayBuffer().then((buffer) => resolve(new Uint8Array(buffer)), reject);
    }, 'image/png');
  });

const drawSvgLayer = async (
  context: CanvasRenderingContext2D,
  svg: SVGSVGElement | undefined,
  outputWidth: number,
  outputHeight: number,
): Promise<void> => {
  if (!svg) return;
  const bounds = svg.getBoundingClientRect();
  if (bounds.width <= 0 || bounds.height <= 0) return;
  const clone = svg.cloneNode(true) as SVGSVGElement;
  clone.setAttribute('width', String(bounds.width));
  clone.setAttribute('height', String(bounds.height));
  clone.setAttribute('viewBox', `0 0 ${bounds.width} ${bounds.height}`);
  clone.style.width = `${bounds.width}px`;
  clone.style.height = `${bounds.height}px`;
  const blob = new Blob([new XMLSerializer().serializeToString(clone)], {
    type: 'image/svg+xml;charset=utf-8',
  });
  const url = URL.createObjectURL(blob);
  try {
    const image = new Image();
    await new Promise<void>((resolve, reject) => {
      image.onload = () => resolve();
      image.onerror = () => reject(new Error('The measurement overlay could not be captured.'));
      image.src = url;
    });
    context.drawImage(image, 0, 0, outputWidth, outputHeight);
  } finally {
    URL.revokeObjectURL(url);
  }
};

const drawLabel = (
  context: CanvasRenderingContext2D,
  value: string,
  x: number,
  y: number,
  align: CanvasTextAlign,
): void => {
  context.textAlign = align;
  context.strokeText(value, x, y);
  context.fillText(value, x, y);
};

const composeKeyImage = async (
  sourceCanvas: HTMLCanvasElement,
  annotationSvg: SVGSVGElement | undefined,
  orientationLabels: PatientOrientationLabels | undefined,
  footer:
    | { kind: 'longitudinal'; viewportRole: 'baseline' | 'followup' }
    | { kind: 'consultation'; selectionSlot: ConsultationSelectionSlot },
  stackPosition: number,
  stackCount: number,
): Promise<HTMLCanvasElement> => {
  if (sourceCanvas.width <= 0 || sourceCanvas.height <= 0) {
    throw new Error('The selected image has not finished rendering.');
  }
  const scale = Math.max(1, sourceCanvas.width / Math.max(1, sourceCanvas.clientWidth));
  const footerHeight = Math.max(48, Math.round(48 * scale));
  const output = document.createElement('canvas');
  output.width = sourceCanvas.width;
  output.height = sourceCanvas.height + footerHeight;
  const context = output.getContext('2d');
  if (!context) throw new Error('The browser could not create the key-image canvas.');

  context.fillStyle = '#000';
  context.fillRect(0, 0, output.width, output.height);
  context.drawImage(sourceCanvas, 0, 0, sourceCanvas.width, sourceCanvas.height);
  await drawSvgLayer(context, annotationSvg, sourceCanvas.width, sourceCanvas.height);

  if (orientationLabels) {
    const inset = 10 * scale;
    context.save();
    context.font = `600 ${Math.max(12, 12 * scale)}px ui-monospace, monospace`;
    context.fillStyle = '#fff';
    context.strokeStyle = '#000';
    context.lineWidth = Math.max(3, 3 * scale);
    context.textBaseline = 'middle';
    drawLabel(context, orientationLabels.left, inset, sourceCanvas.height / 2, 'left');
    drawLabel(
      context,
      orientationLabels.right,
      sourceCanvas.width - inset,
      sourceCanvas.height / 2,
      'right',
    );
    context.textBaseline = 'top';
    drawLabel(context, orientationLabels.top, sourceCanvas.width / 2, inset, 'center');
    context.textBaseline = 'bottom';
    drawLabel(
      context,
      orientationLabels.bottom,
      sourceCanvas.width / 2,
      sourceCanvas.height - inset,
      'center',
    );
    context.restore();
  }

  context.fillStyle = '#071411';
  context.fillRect(0, sourceCanvas.height, output.width, footerHeight);
  context.fillStyle = '#e7f3ef';
  context.textAlign = 'center';
  context.textBaseline = 'middle';
  context.font = `600 ${Math.max(10, 11 * scale)}px ui-monospace, monospace`;
  context.fillText(
    footer.kind === 'consultation'
      ? 'UNREVIEWED · CONSULTATION REFERENCE VIEW · NOT FOR DIAGNOSIS'
      : 'UNREVIEWED · DERIVED DISPLAY KEY IMAGE · NOT FOR DIAGNOSIS',
    output.width / 2,
    sourceCanvas.height + footerHeight * 0.34,
  );
  context.fillStyle = '#9bb0aa';
  context.font = `400 ${Math.max(9, 9 * scale)}px ui-monospace, monospace`;
  context.fillText(
    footer.kind === 'consultation'
      ? `${footer.selectionSlot === 'view_a' ? 'VIEW A' : 'VIEW B'} · slice ${stackPosition} / ${stackCount} · no chronology or response relationship`
      : `${footer.viewportRole.toUpperCase()} · slice ${stackPosition} / ${stackCount} · original DICOM retained separately`,
    output.width / 2,
    sourceCanvas.height + footerHeight * 0.72,
  );
  return output;
};

export const scopeMeasurementPacketToInstance = (
  packet: MeasurementEvidencePacket,
  seriesId: string,
  instanceId: string,
  createdAt: string,
): MeasurementEvidencePacket => ({
  schema_version: '3.0.0',
  created_at: createdAt,
  review_status: 'unreviewed',
  measurements: packet.measurements.filter(
    (measurement) =>
      measurement.source.series_id === seriesId && measurement.source.instance_id === instanceId,
  ),
  limitations: [
    ...packet.limitations,
    'This measurement packet is scoped to the single displayed source instance in the accompanying key image.',
  ],
});

export const buildKeyImageEvidencePacket = async ({
  createdAt,
  source,
  display,
  imageWidth,
  imageHeight,
  pngBytes,
  measurementPacket,
  measurementBytes,
}: BuildPacketInput): Promise<KeyImageEvidencePacket> => ({
  schema_version: '2.0.0',
  created_at: createdAt,
  review_status: 'unreviewed',
  artifact_type: 'derived_display_key_image',
  source,
  display,
  image: {
    filename: 'key-image.png',
    mime_type: 'image/png',
    width_px: imageWidth,
    height_px: imageHeight,
    sha256: await sha256Hex(pngBytes),
  },
  measurement_evidence: {
    filename: 'measurements.json',
    schema_version: '3.0.0',
    measurement_count: measurementPacket.measurements.length,
    tracking_ids: measurementPacket.measurements.map((measurement) => measurement.tracking_id),
    sha256: await sha256Hex(measurementBytes),
  },
  implementation: {
    name: 'ScanView key-image exporter',
    version: '0.2.0',
    renderer: 'Cornerstone3D 5.8.2',
  },
  limitations: [
    'This is a display-rendered, unreviewed derivative; the original DICOM instance remains authoritative.',
    'Window/level, zoom, pan, annotations, and orientation labels reflect the captured viewer state.',
    'Measurements and annotations require qualified clinician review and do not establish diagnosis or treatment response.',
    'DICOM pixels may contain burned-in identifiers or recognizable anatomy; treat the entire archive as sensitive medical data.',
  ],
});

export const buildConsultationKeyImageEvidencePacket = async ({
  createdAt,
  source,
  display,
  imageWidth,
  imageHeight,
  pngBytes,
  measurementPacket,
  measurementBytes,
}: BuildConsultationPacketInput): Promise<ConsultationKeyImageEvidencePacket> => ({
  schema_version: '1.0.0',
  created_at: createdAt,
  review_status: 'unreviewed',
  artifact_type: 'derived_display_consultation_key_image',
  source,
  display,
  image: {
    filename: 'key-image.png',
    mime_type: 'image/png',
    width_px: imageWidth,
    height_px: imageHeight,
    sha256: await sha256Hex(pngBytes),
  },
  measurement_evidence: {
    filename: 'measurements.json',
    schema_version: '3.0.0',
    measurement_count: measurementPacket.measurements.length,
    tracking_ids: measurementPacket.measurements.map((measurement) => measurement.tracking_id),
    sha256: await sha256Hex(measurementBytes),
  },
  implementation: CONSULTATION_KEY_IMAGE_IMPLEMENTATION,
  limitations: [...CONSULTATION_KEY_IMAGE_LIMITATIONS],
});

export const createKeyImageArchive = async ({
  viewportCanvas,
  annotationSvg,
  orientationLabels,
  viewportRole,
  source,
  display,
  measurementPacket,
  createdAt = new Date().toISOString(),
}: KeyImageArchiveInput): Promise<KeyImageArchive> => {
  const scopedMeasurements = scopeMeasurementPacketToInstance(
    measurementPacket,
    source.series_id,
    source.instance_id,
    createdAt,
  );
  const measurementBytes = strToU8(`${JSON.stringify(scopedMeasurements, null, 2)}\n`);
  const composite = await composeKeyImage(
    viewportCanvas,
    annotationSvg,
    orientationLabels,
    { kind: 'longitudinal', viewportRole },
    display.stack_position,
    display.stack_count,
  );
  const pngBytes = await canvasToPngBytes(composite);
  const packet = await buildKeyImageEvidencePacket({
    createdAt,
    source,
    display: {
      viewport_role: viewportRole,
      ...display,
      viewport_width_px: viewportCanvas.width,
      viewport_height_px: viewportCanvas.height,
      patient_orientation: orientationLabels,
    },
    imageWidth: composite.width,
    imageHeight: composite.height,
    pngBytes,
    measurementPacket: scopedMeasurements,
    measurementBytes,
  });
  const packetBytes = strToU8(`${JSON.stringify(packet, null, 2)}\n`);
  const archive = zipSync(
    {
      'key-image.json': [packetBytes, { level: 6 }],
      'key-image.png': [pngBytes, { level: 6 }],
      'measurements.json': [measurementBytes, { level: 6 }],
    },
    { level: 6 },
  );
  const timestamp = createdAt.replace(/[-:]/g, '').replace(/\.\d{3}Z$/, 'Z');
  const filename = `scanview-key-image-${timestamp}-${viewportRole}.zip`;
  return { filename, packet, bytes: archive };
};

export const createConsultationKeyImageArchive = async ({
  viewportCanvas,
  annotationSvg,
  orientationLabels,
  selectionSlot,
  source,
  display,
  measurementPacket,
  createdAt = new Date().toISOString(),
}: ConsultationKeyImageArchiveInput): Promise<ConsultationKeyImageArchive> => {
  const scopedMeasurements = scopeMeasurementPacketToInstance(
    measurementPacket,
    source.series_id,
    source.instance_id,
    createdAt,
  );
  const measurementBytes = strToU8(`${JSON.stringify(scopedMeasurements, null, 2)}\n`);
  const composite = await composeKeyImage(
    viewportCanvas,
    annotationSvg,
    orientationLabels,
    { kind: 'consultation', selectionSlot },
    display.stack_position,
    display.stack_count,
  );
  const pngBytes = await canvasToPngBytes(composite);
  const packet = await buildConsultationKeyImageEvidencePacket({
    createdAt,
    source,
    display: {
      selection_slot: selectionSlot,
      ...display,
      viewport_width_px: viewportCanvas.width,
      viewport_height_px: viewportCanvas.height,
      patient_orientation: orientationLabels,
    },
    imageWidth: composite.width,
    imageHeight: composite.height,
    pngBytes,
    measurementPacket: scopedMeasurements,
    measurementBytes,
  });
  const packetBytes = strToU8(`${JSON.stringify(packet, null, 2)}\n`);
  const archive = zipSync(
    {
      'key-image.json': [packetBytes, { level: 6 }],
      'key-image.png': [pngBytes, { level: 6 }],
      'measurements.json': [measurementBytes, { level: 6 }],
    },
    { level: 6 },
  );
  const timestamp = createdAt.replace(/[-:]/g, '').replace(/\.\d{3}Z$/, 'Z');
  const slot = selectionSlot.replace('_', '-');
  return {
    filename: `scanview-consultation-key-image-${timestamp}-${slot}.zip`,
    packet,
    bytes: archive,
  };
};

export const downloadArchive = (bytes: Uint8Array, filename: string): void => {
  const archiveBuffer = bytes.buffer.slice(
    bytes.byteOffset,
    bytes.byteOffset + bytes.byteLength,
  ) as ArrayBuffer;
  const url = URL.createObjectURL(new Blob([archiveBuffer], { type: 'application/zip' }));
  const link = document.createElement('a');
  link.href = url;
  link.download = filename;
  link.style.display = 'none';
  document.body.append(link);
  link.click();
  link.remove();
  window.setTimeout(() => URL.revokeObjectURL(url), 0);
};

export const exportKeyImageArchive = async (
  input: KeyImageArchiveInput,
): Promise<KeyImageArchive> => {
  const result = await createKeyImageArchive(input);
  downloadArchive(result.bytes, result.filename);
  return result;
};
